support for not all DUT ports connected to a fanout switch

sanmalho-git opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [x] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Approach
#### What is the motivation for this PR?
Currently, it is required that all ports on DUT are in use and are connected to a fanout. However, there is a need to be able to run tests where all ports are not in use. Specifically, when dealing with
- New Sonic devices
- Sonic device with a front panel port used for in-band management
- Chassis with lots of ports and multiple asics, were every port on every asic is not required to be covered
- Chassis as a DUT, where the number of ports can be in hundreds
- Expensive, high speed ports like 400G (hard to go from 400G down to 1/10G)

So, need to add support where not all DUT ports are connected to fanout and are thus not part of the testing.

#### How did you do it?

**conn_graph_facts:**
  
conn_graph_facts.py returns device_vlan_map_list. This used to be a dictionary with key being hostname and value being a list of vlanids for the fanout ports. Have modified this where the value instead of being a list of vlanids, it is a dictionary with key being the port_index and the value being the vlan id. This port_index is what gets put in the topology file. We get the port by looking at the host_port_vlans defined in the conn_graph for that device. This host_port_vlans has key being the Ethernet port (like Ethernet10) and value being a dictionary with key 'vlanlist' being a the list of fanout vlans. We check against all the ports 'vlanlist' to get the port on the DUT that connects to this fanout vlan, and then split on 'Ethernet' to get the port index.

For example - lets say on dut with hostname "dut1", we have port Ethernet10 connected to fanout w/ vlan 120, and Ethernet11 connected to fanout w/ vlan 121, then we would have:
```
  "host_port_vlans":
    {
      "Ethernet10": {
         "mode": "Access",
         "vlanids": "120",
         "vlanlist": [
            120
         ]
      },
      "Ethernet11": {
         "mode": "Access",
         "vlanids": "121",
         "vlanlist": [
            121
         ]
      }
   }

  "VlanList": [
     120,
     121
  ]
```
For vlan 120 in VlanList, we would iterate through host_port_vlans to find the port that has vlan 120 - in our case "Ethernet10". The port_index would then be "10". Similarly, for vlan 121, the port_index would be "Ethernet11".

So, returned device_vlan_map_list would be:
```
    "dut1" : {
       "10" : 120,
       "11" : 121
    }
```

**vlan_port/kvm_port/mellanox_simx_port:**
   - Updated to return (dut_fp_ports) a dictionary with key being the port index (same as in the topo file) and vlan being the port - instead of the just the list of ports.

**bind/unbind vm_topology:**
   - vlan_index is now a string in the dictionary of dut_fp_ports
   - updated regexp for checking valid vlan for multi-dut to be of the format '<dutIndex>.<portIndex>@<ptfIndex>'

**remove_dut_port.yml (bug fix):**
  - set cmd to "remove" instead of "create" in vlan_port module call.
                                                                              
#### How did you verify/test it?

Tested against pizza box DUT with all DUT ports connected to a fanout, and also against another DUT where we have only 4 of the 52 ports connected to a different fanout.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

sanmalho-git commented:
@lolyu can you please review these changes.

lolyu commented on ansible/library/conn_graph_facts.py:
```diff
@@ -338,7 +338,19 @@ def main():
+                    found_port_for_vlan = False
+                    for a_port in port_vlans:
+                        if a_host_vlan in port_vlans[a_port]['vlanlist']:
+                            port_index = a_port.split("Ethernet")[1]
```
`port_index` is actually the interface suffix number, right? Should we use the actual indexes(like 0, 1, 2 ...)?

sanmalho-git replied:
I am not following your question.

I am trying to get the interface suffix using split based on 'Ethernet'.  For the example that I gave above in the PR description,  we would have the following:

```
port_vlans = {
      "Ethernet10": {
         "mode": "Access",
         "vlanids": "120",
         "vlanlist": [
            120
         ]
      },
      "Ethernet11": {
         "mode": "Access",
         "vlanids": "121",
         "vlanlist": [
            121
         ]
      }
   }

host_vlan["VlanList"] = [120, 121]
```

What gets put into the topology file (host_interfaces or vlans under VMs) for the ports Ethernet10 and Ethernet11 are '10' and '11' respectively.

So, we want to match port suffix '10' to vlan '120', and port suffix '11' to vlan '121'.

lolyu replied:
Yes, what if the interfaces are like `Ethernet0`, `Ethernet2`, `Ethernet4`(the suffix numbers are not consecutive).

sanmalho-git replied:
In this case, wouldn't we have '0', '2' and '4' in the topology file, instead of 0, 1, 2?

And the returned device_vlan_map_list would be:
```
"dut1": {
   "0" : 120,
   "2", 121
}
```

sanmalho-git replied:
You are correct - this approach would not work for DUT's that don't have consecutive ports. 

One option that I can see to solve this that we add something similar to port_alias_to_name_map in minigraph_facts to conn_graph_facts which gives the lists of ports based on hwsku.

Would it be ok to do it this way - if so I can code it up.

lolyu replied:
Yes, it will be much more better as indexes are consecutive. With this change, I believe there will be errors in `add-topo`, will you fix those plays?

sanmalho-git replied:
I have committed the code - can you please review again.

lolyu replied:
Making `conn_graph_facts` the ability to get port index from alias seems to introduce code duplication in this way. How about simply make `device_vlan_map_list` simply be like
```
    "dut1" : {
       "Ethernet40" : 120,
       "Ethernet44" : 121
    }
```

sanmalho-git replied:
If we do this, then the problem shifts to generation of 'dut_fp_ports' - specifically in roles/vm_set/library/vlan_port.py. dut_fp_ports eventually is passed into vm_topology where we need indices.

I have another PR to introduce common module_utils - https://github.com/Azure/sonic-mgmt/pull/2623/.  The idea was to put this common code between conn_graph_facts and minigraph_facts into the module_utils.  Once that PR is approved, then would move this common code into module_utils and then push this PR.

lolyu commented on ansible/library/conn_graph_facts.py:
```diff
@@ -338,7 +466,25 @@ def main():
+                for a_host_vlan in host_vlan["VlanList"]:
+                    # Get the corresponding port for this vlan from the port vlan list for this hostname
+                    found_port_for_vlan = False
+                    for a_port in port_vlans:
```
Do we really need to check the vlan ids in `VlanList` is in `port_vlans`? `VlanList`  comes from the combination of all the vlan lists defined in `self.vlanport[hostname]['vlanlist']`, while `port_vlans` is just `self.vlanport`?

sanmalho-git replied:
I think it is safe to assume that self.vlanport[hostname]['vlanlist'] would always be a list of 1.  Can you please confirm? If so, then I can change the code appropriately.

lolyu reviewed (approved):
Please fix the conflict.

sanmalho-git commented:
@lolyu - I have re-based and fixed the merge conflict. Also, the module_utils PR has been merged into master. So, moved the common code of port_alias_to_name_map between conn_graph_facts and minigraph_facts to module_utils/port_utils.py.

Can you please give it a final review before we merge this PR into master.

lolyu commented:
Retest this please

lolyu commented on ansible/library/conn_graph_facts.py:
```diff
@@ -287,6 +290,23 @@ def find_graph(hostnames):
+    port_alias_to_name_map = get_port_alias_to_name_map(hwsku)
+
+    # Create inverse mapping between port name and alias
+    port_name_to_alias_map = {v: k for k, v in port_alias_to_name_map.iteritems()}
```
Could simply get the port name list with
```python
port_name_list = port_alias_to_name_map.values()
```

lolyu commented on ansible/roles/vm_set/library/vlan_port.py:
```diff
@@ -150,12 +144,13 @@ def main():
+        fp_ports = {}
         for dut_vlans in vlan_ids:
             dut_vlans.sort()
             fp_ports.append([fp_port_templ % vid for vid in dut_vlans])
```
Should we ensure the same behavior regarding multiple-DUTs' testbed?

sanmalho-git replied:
I am not sure if is_multi_duts is still being used. All calls to vlan_port from ansible/roles/vm_set do not pass is_multi_duts and are thus false.  But, I am not aware of how this is being used by SpyTest. If they don't use it, then we can remove support multi_dut from vlan_port as well

lolyu replied:
Looks like the `vlan_port` will never be called with the nested `vlan_ids` since it is called with a loop.

sanmalho-git replied:
should we remove is_multi_duts support from vlan_port then.

lolyu commented on ansible/roles/vm_set/library/vm_topology.py:
```diff
@@ -653,7 +653,7 @@ def add_host_ports(self):
                     self.add_dut_if_to_docker(ptf_if, fp_port)
             else:
-                fp_port = self.duts_fp_ports[self.duts_name[0]][intf]
+                fp_port = self.duts_fp_ports[self.duts_name[0]][str(intf)]
```
In `add_host_ports`, the type change from list to dictionary will break the code in the circumstances that `self._is_multi_duts` evaluates to `True`.

sanmalho-git replied:
Good catch - my testing was dealing with no host interfaces in a multi-dut environment and thus did not catch this. Committed a fix for this

lolyu reviewed (approved):
LGTM

lolyu commented:
Thanks, @sanmalho-git.
